Source/WebCore/page/text-extraction/TextExtraction.cpp

```diff
@@ -1773,5 +1784,105 @@ std::optional<SimpleRange> rangeForExtractedText(const LocalFrame& frame, Extrac
+{
+    return WTFMove(data).map([](auto&& data) -> FilterRule {
+        auto&& [name, urlPattern, scriptSource] = WTFMove(data);
+        if (urlPattern.isEmpty())
+            return { WTFMove(name), std::nullopt, WTFMove(scriptSource) };
+
+        auto regex = Yarr::RegularExpression { urlPattern, { Yarr::Flags::IgnoreCase } };
+        if (!regex.isValid())
+            return { WTFMove(name), std::nullopt, WTFMove(scriptSource) };
+
+        return { WTFMove(name), { WTFMove(regex) }, WTFMove(scriptSource) };
+    });
+}
+
+static DOMWrapperWorld& filteringWorld()
+{
+    static NeverDestroyed<RefPtr<DOMWrapperWorld>> world;
+    static std::once_flag onceFlag;
+    std::call_once(onceFlag, [&] {
+        world.get() = DOMWrapperWorld::create(commonVM(), DOMWrapperWorld::Type::Internal, "Text Extraction Filtering Rules"_s);
+    });
+    return *world.get();
+}
+
+void applyRules(const String& input, std::optional<NodeIdentifier>&& containerNodeID, const Vector<FilterRule>& rules, Page& page, CompletionHandler<void(const String&)>&& completion)
+{
+    if (rules.isEmpty())
+        return completion(input);
+
+    RefPtr mainFrame = page.localMainFrame();
+    if (!mainFrame)
+        return completion(input);
+
+    RefPtr document = mainFrame->document();
+    if (!document)
+        return completion(input);
+
+    RefPtr containerNode = resolveNodeWithBodyAsFallback(page, WTFMove(containerNodeID));
+    if (!containerNode)
+        return completion(input);
+
+    Ref world = filteringWorld();
+    auto arguments = [&] {
+        ArgumentMap argumentMap;
+        argumentMap.reserveInitialCapacity(2);
+        argumentMap.add("input"_s, [input](auto& lexicalGlobalObject) {
+            JSLockHolder lock { &lexicalGlobalObject };
+            return JSValue { jsString(commonVM(), input) };
+        });
+        argumentMap.add("containerNode"_s, [containerNode, mainFrame, world = world.copyRef()](auto& lexicalGlobalObject) {
+            if (!containerNode)
+                return jsNull();
+
+            JSLockHolder lock { &lexicalGlobalObject };
+            return toJS(&lexicalGlobalObject, mainFrame->checkedScript()->globalObject(world), *containerNode);
+        });
+        return std::make_optional(WTFMove(argumentMap));
+    }();
+
+    auto filteredStrings = Box<Vector<String>>::create();
+    auto aggregator = MainRunLoopCallbackAggregator::create([completion = WTFMove(completion), input, filteredStrings] mutable {
+        if (filteredStrings->isEmpty())
+            return completion(input);
+
+        auto shortestFilteredString = std::ranges::min(*filteredStrings, { }, [](auto& string) {
+            return string.length();
+        });
+        completion(WTFMove(shortestFilteredString));
+    });
+
+    auto urlString = document->url().string();
+    for (auto& [name, urlPattern, source] : rules) {
+        if (urlPattern && urlPattern->match(urlString) < 0)
```

Comment:
We also want to skip the iteration when `urlPattern == std::nullopt`, right? something like this?
```suggestion
        if (!urlPattern || urlPattern->match(urlString) < 0)
```

Reply:
or maybe `urlPattern == std::nullopt` means allow any URL pattern.

Reply:
Ah, that's right — optional is intended to default to "apply regardless". I can make this more explicit by making it a `Variant` instead.